opennms-webapp/src/main/java/org/opennms/web/security/JSessionIdNoCacheFilter.java

```diff
@@ -0,0 +1,88 @@
+ * along with OpenNMS(R).  If not, see:
+ *      http://www.gnu.org/licenses/
+ *
+ * For more information contact:
+ *     OpenNMS(R) Licensing <license@opennms.org>
+ *     http://www.opennms.org/
+ *     http://www.opennms.com/
+ *******************************************************************************/
+
+package org.opennms.web.security;
+
+import java.io.IOException;
+import java.util.Collection;
+
+import javax.servlet.Filter;
+import javax.servlet.FilterChain;
+import javax.servlet.FilterConfig;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.http.HttpServletResponse;
+
+/** Sets the header "cache-control: no-cache" if the response contains the jsessionid. This is done for security reasons;
+ * Web servers use session cookies to identify the active user sessions.  Disclosing these session cookies to an
+ * attacker can result in a session hijacking attack. With this in mind, session cookies should be treated as sensitive
+ * data and should be well protected. With a secure cache-control policy in place, session cookies are typically stored
+ * in the browser memory instead of flushed to the hard drive. When a browsing session is terminated, the corresponding
+ * session cookies are deleted from the client machine.  However, session cookies can also be kept in a cached web page.
+ * For example, when a user login in a web site, the returned web page may contain session cookies in the “set-cookie”
+ * response headers.  If the “cache-control” header in this page is defined as “public”, all proxy servers and gateways
+ * between the client and server are allowed to cache this page. Thus, the risk of exposing these sensitive session
+ * cookies to an attacker is significantly increased. It’s recommended to change the cache-control header to secure
+ * values, e.g., no-cache.    Similar to a password, if an attacker steals the session cookies that represent a valid
+ * user session, they can then use them to masquerade as victim users and access their personal data. If a victim user
+ * has administrative privileges, then the security of entire web site is at risk.
+ */
+public class JSessionIdNoCacheFilter implements Filter {
+
+    private final static String DEFAULT_SESSION_ID_NAME = "JSESSIONID";
+
+    @Override
+    public void doFilter(ServletRequest request, ServletResponse response, FilterChain filterChain) throws IOException, ServletException {
+
+        filterChain.doFilter(request, response);
+        if(response instanceof HttpServletResponse) {
+            addHeaderIfRequired((HttpServletResponse)response);
+        }
+    }
+
+    private void addHeaderIfRequired(HttpServletResponse response) {
+        Collection<String> cookies = response.getHeaders("Set-Cookie");
+        for( String cookie : cookies) {
+            if(cookie.contains(DEFAULT_SESSION_ID_NAME)){
+                response.addHeader("Cache-Control", "no-cache");
```

Comment:
@mvrueden I add here the Cache-Control header. I can verify that it is set via debugging but it never reaches the client. I debugged all the way up through the Jetty code but I couldn't find a place where it is lost. Do you have an idea why it gets lost and or another idea how to fix this security vulnerability? Thanks!

Reply:
Filters are processed in the order of the `filter-mapping` definitions. I gave this a try and moved the new `filter-mapping` to be the last one:
```
HTTP/1.1 200 OK
Date: Mon, 09 Sep 2019 17:23:30 GMT
X-Frame-Options: SAMEORIGIN
Content-Type: text/html;charset=utf-8
Set-Cookie: JSESSIONID=node0w4cziqnkjair1gkn4ojtsquwi78.node0;Path=/opennms
Expires: Thu, 01 Jan 1970 00:00:00 GMT
Cache-Control: no-cache
Content-Length: 5519
Server: Jetty(9.4.18.v20190429)
```

Reply:
was my mistake - a caching problem...